In [228]:
import pandas as pd

# Работа с Pandas

Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [229]:
df = pd.read_csv('googleplaystore.csv') # ваш путь к файлу с данным

Отобразите количество строк и столбцов в DataFrame

In [230]:
df.shape

(10841, 13)

# PRO часть (Не обязательна к выполнению, но желательна :))


Измените названия всех столбцов приведя их к нижнему регистру

In [231]:
#Приводим название столбцов к соответствию по pep8 (было сказанно в уроке)
df.columns = df.columns.str.lower().str.replace(' ', '_')

Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [232]:
df.describe(include=object)

app category reviews                size    installs   type  \
count    10841    10841   10841               10841       10841  10840   
unique    9660       34    6002                 462          22      3   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free   
freq         9     1972     596                1695        1579  10039   

       content_rating genres    last_updated         current_ver android_ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

1) Из данных видно, что датафрейм содержит 10841 строк. При этом столбцы type, content_rating, current_ver, androind_ver содержат пропуски.

2) В датасете 9660 уникальных названий приложений в 34 категориях.

3) Чаще всего встречаются приложения в категории FAMILY, также приложения в большенстве своем бесплатные, для версии андройд 4.1 и выше.

In [233]:
df.describe()

rating         price
count  9367.000000  10841.000000
mean      4.193338      1.027273
std       0.537431     15.948971
min       1.000000      0.000000
25%       4.000000      0.000000
50%       4.300000      0.000000
75%       4.500000      0.000000
max      19.000000    400.000000

Видно что максимальное значение рейтинга равно 19, что является ошибкой и нам нужно это поправить

In [234]:
max_index = df.rating.idxmax()
# Удаляем строку с максимальным значением
df.drop(max_index, axis=0, inplace=True)

# Выводим метрики датафрейма без строки с максимальным значением
df.describe()

rating         price
count  9366.000000  10840.000000
mean      4.191757      1.027368
std       0.515219     15.949703
min       1.000000      0.000000
25%       4.000000      0.000000
50%       4.300000      0.000000
75%       4.500000      0.000000
max       5.000000    400.000000

Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [235]:
display(df.query('category == "ART_AND_DESIGN" & rating > 4.5'))

app        category  \
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN   
9                         Kids Paint Free - Drawing Fun  ART_AND_DESIGN   
13                                Mandala Coloring Book  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
26                        Colorfit - Drawing & Coloring  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
36       UNICORN - Color By Number & Pixel Art Coloring  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   

      rating reviews  size     installs  type  price content_rating  \
2        4.7   87510  8.7M   5,000,000+  Free   0.00       Everyone   
9        4.7     121  3.1M      10,000+  Free   0.00       Everyone   
13       4.6    4326   21M     100,000+  Free   0.00       Everyone   
16       4.7    3632  5.5M     500,000+  Free   0.00       Everyone   
19       4.6  224399   31M  10,000,000+  Free   0.00       Everyone   
22       4.7    7699  4.2M     500,000+  Free   0.00   Everyone 10+   
24       4.7     118   23M      50,000+  Free   0.00       Everyone   
25       4.8     192  6.0M      10,000+  Free   0.00       Everyone   
26       4.7   20260   25M     500,000+  Free   0.00       Everyone   
34       4.7     353  4.2M      10,000+  Free   0.00           Teen   
36       4.7    8145   24M     500,000+  Free   0.00       Everyone   
38       4.7     158   11M      10,000+  Free   0.00       Everyone   
45       4.7  174531   24M  10,000,000+  Free   0.00       Everyone   
46       4.6    1070   26M     100,000+  Free   0.00       Everyone   
3982     4.8     253  3.7M      50,000+  Free   0.00       Everyone   
4749     4.7    5754  4.4M     100,000+  Free   0.00       Everyone   
4755     4.7     801  3.5M       5,000+  Paid   1.99       Everyone   
4759     4.8    1216  8.6M      10,000+  Paid   1.99       Everyone   
4764     4.7     149  3.5M       1,000+  Paid   1.99       Everyone   
5254     4.8      48   25M       5,000+  Free   0.00       Everyone   
6930     4.8      41  1.9M       5,000+  Free   0.00       Everyone   
8888     5.0       1  2.9M         100+  Free   0.00       Everyone   

                       genres       last_updated current_ver   android_ver  
2                Art & Design     August 1, 2018       1.2.4  4.0.3 and up  
9     Art & Design;Creativity       July 3, 2018         2.8  4.0.3 and up  
13               Art & Design      June 26, 2018       1.0.4    4.4 and up  
16               Art & Design      July 31, 2018         3.1    4.1 and up  
19               Art & Design      July 30, 2018       5.5.4    4.1 and up  
22               Art & Design      July 12, 2018     2.2.6.2  4.0.3 and up  
24               Art & Design       July 7, 2018       1.1.3    4.1 and up  
25               Art & Design     April 25, 2018         1.5    3.0 and up  

Посчитатайте какое количество приложений содержится в рейтенге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.

In [236]:
df[df.category =='GAME'].rating.value_counts(bins=100).head(10)

(4.48, 4.52]    174
(4.36, 4.4]     166
(4.28, 4.32]    151
(4.56, 4.6]     121
(4.16, 4.2]     115
(4.08, 4.12]     72
(4.68, 4.72]     67
(3.96, 4.0]      55
(3.76, 3.8]      41
(3.88, 3.92]     35
Name: rating, dtype: int64

Из вывода видно что 174 приложения категории GAME имеют рейтинг от 4.48 до 4.52

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [237]:
# Создание функции для категоризации рейтинга
def categorize_rating(rating):
    if rating >= 4.5:
        return 'High rating'
    elif rating >= 3.8:
        return 'Middle rating'
    else:
        return 'Low rating'

# Применение функции к столбцу рейтинга
df['categorical_rating'] = df['rating'].apply(categorize_rating)

# Вывод результата
df.head(5)

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [238]:
print(df.nunique().sort_values(ascending=False))

app                   9659
reviews               6001
current_ver           2831
last_updated          1377
size                   461
genres                 119
price                   92
rating                  39
category                33
android_ver             33
installs                21
content_rating           6
categorical_rating       3
type                     2
dtype: int64


__Super-Hard задания__

Решите прошлую задачу, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [239]:
# Применение lambda функции к столбцу рейтинга
df['categorical_rating'] = df['rating'].apply(lambda x: 'High rating' if x >= 4.5 else ('Middle rating' if x >= 3.8 else 'Low rating'))
# Вывод результата
df.head(5)

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [240]:
df['mean_cat_rating'] = df.groupby('category')['rating'].transform('mean')
# Выведем 10 случайных строк 
df.sample(10)

app            category  \
8481                     Visuelles Wörterbuch Audio-App              FAMILY   
2340                      Pill Identifier and Drug list             MEDICAL   
8185                              Geometry Dash SubZero                GAME   
2938   PicsArt Photo Studio: Collage Maker & Pic Editor         PHOTOGRAPHY   
6770                                    BT Communicator              SOCIAL   
8803                            DR Sebi's Alkaline List              FAMILY   
10005                              European War 6: 1804              FAMILY   
7829                                       Mobile CS:GO                GAME   
5782              AXE.IO - Brutal Survival Battleground                GAME   
9295                                           EF Coach  HEALTH_AND_FITNESS   

       rating  reviews  size      installs  type  price content_rating  \
8481      3.1      192   66M       10,000+  Free   0.00       Everyone   
2340      4.0      488   17M      100,000+  Free   0.00       Everyone   
8185      4.6   260527   49M   10,000,000+  Free   0.00       Everyone   
2938      4.5  7594559   34M  100,000,000+  Free   0.00           Teen   
6770      3.8       32  1.6M       10,000+  Free   0.00           Teen   
8803      4.0       68  5.5M        5,000+  Free   0.00       Everyone   
10005     4.2     1330   96M       10,000+  Paid   0.99           Teen   
7829      3.1     1015  643k      100,000+  Free   0.00       Everyone   
5782      4.2     3845   76M      500,000+  Free   0.00   Everyone 10+   
9295      4.8       17   14M          500+  Free   0.00       Everyone   

                 genres       last_updated current_ver   android_ver  \
8481          Education       July 7, 2017       1.1.6    4.1 and up   
2340            Medical     August 1, 2018         3.8    4.1 and up   
8185             Arcade  December 21, 2017        1.00    4.0 and up   
2938        Photography     August 6, 2018      9.40.3  4.0.3 and up   
6770             Social  December 16, 2016         1.0    1.6 and up   
8803          Education      April 5, 2017         0.1    2.3 and up   
10005          Strategy      July 18, 2018       1.1.0    4.4 and up   
7829             Action  November 11, 2015         1.3  2.3.3 and up   
5782             Arcade     August 3, 2018       1.6.1    4.1 and up   
9295   Health & Fitness     March 12, 2018       1.1.1    4.4 and up   

      categorical_rating  mean_cat_rating  
8481          Low rating         4.192272  
2340       Middle rating         4.189143  
8185         High rating         4.286326  
2938         High rating         4.192114  
6770       Middle rating         4.255598  
8803       Middle rating         4.192272  
10005      Middle rating         4.192272  
7829          Low rating         4.286326  
5782       Middle rating         4.286326  
9295         High rating         4.277104

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого стркоу с названием `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

In [241]:
def get_month(date_str):
    try:
        date = pd.to_datetime(date_str)
        month = date.month
        return month
    except:
        return 'miss_date'

In [242]:
# Применяем функцию к столбцу last updated и создаем новый столбец day_of_update
df['day_of_update'] = df['last_updated'].apply(get_month)
#Выводим 5 случайных строк
df.sample(5)

app         category  rating  reviews  \
1845                        Fallout Shelter             GAME     4.6  2719142   
343                                My Tele2    COMMUNICATION     4.3   158679   
4479  G-Pix [Android P] Dark EMUI 8/5 THEME  PERSONALIZATION     4.1      169   
5428                       Legion of Heroes           FAMILY     4.3   143087   
4041                           Temple Run 2             GAME     4.3  8116142   

      size      installs  type  price content_rating           genres  \
1845   25M   10,000,000+  Free    0.0           Teen       Simulation   
343   8.8M    5,000,000+  Free    0.0       Everyone    Communication   
4479   24M       10,000+  Free    0.0       Everyone  Personalization   
5428   69M    1,000,000+  Free    0.0   Everyone 10+     Role Playing   
4041   62M  500,000,000+  Free    0.0       Everyone           Action   

        last_updated current_ver android_ver categorical_rating  \
1845   June 11, 2018     1.13.12  4.1 and up        High rating   
343   August 3, 2018       2.4.1  4.4 and up      Middle rating   
4479  March 25, 2018           7  4.4 and up      Middle rating   
5428   July 12, 2018      1.9.22  4.0 and up      Middle rating   
4041    July 5, 2018      1.49.1  4.0 and up      Middle rating   

      mean_cat_rating  day_of_update  
1845         4.286326              6  
343          4.158537              8  
4479         4.335987              3  
5428         4.192272              7  
4041         4.286326              7